### PR TITLE
[FLINK-6617][table] Improve JAVA and SCALA logical plans consistent test

### PR DESCRIPTION
**Pull request by sunjincheng121**
In this PR. I had make some changes as follows:
1. Add all the tableAPI JAVA and SCALA logical plans consistent test.
2. Add `validation` folder for steam which keep consistent with batch.
3. Refactoring the test structure.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-6617][table] Improve JAVA and SCALA logical plans consistent test")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/validation/GroupWindowAggregationsTest.scala [hunk 1]
```diff
@@ -16,19 +16,19 @@
  * limitations under the License.
  */
 
-package org.apache.flink.table.api.scala.stream.table
+package org.apache.flink.table.api.scala.stream.table.validation
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/validation/GroupWindowAggregationsTest.scala]: I think tests in `validation` packages should all test for a `ValidationException`. Can you split this class?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/TableTestBase.scala [hunk 2]
```diff
@@ -120,7 +125,8 @@ object TableTestUtil {
 }
 
 case class BatchTableTestUtil() extends TableTestUtil {
-
+  val javaEnv = mock(classOf[JExecutionEnvironment])
+  val javaTableEnv = TableEnvironment.getTableEnvironment(javaEnv)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/utils/TableTestBase.scala]: Maybe we can rename the Scala environment to `tableEnv` to be consistent.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/stringexpr/UnionStringExpressionTest.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+package org.apache.flink.table.api.scala.stream.table.stringexpr
+
+import org.apache.flink.api.scala._
+import org.apache.flink.table.api.scala._
+import org.apache.flink.table.utils.TableTestBase
+import org.junit.Test
+
+class UnionStringExpressionTest extends TableTestBase {
+  @Test
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/stringexpr/UnionStringExpressionTest.scala]: Add an empty line here.

**Comment by sunjincheng121** [comment]
Hi @twalthr  Thanks a log for your reviewing. I completely agree with your suggestions. I petty want refactoring the test structure. I have updated the PR. according your comments.

Thanks,
SunJincheng

**Comment by sunjincheng121** [comment]
@twalthr I have rebased code against master.

Best,
SunJincheng

**Comment by sunjincheng121** [comment]
**I once again seriously consider the Refactoring work, according to the following principles updated the PR.**

   **First, the test type and purpose(4 types):**
* Unit Test: to ensure the correctness of the most granular features
* StringExpression Test: Verify the consistency of java and scalaAPI
* Validation Test: Check the API exception information
* Integration Test: Integration testing, verifying runtime and end-to-end correctness

   **Second, the directory structure(folders) and purpose:**
  
    The TEST directory contains all the paths (folders) of the SRC directory - the functional tests included three type test case (**Unit Test**, **StringExpression Test**,**Validation Test**)
   **Note:** runtime directory contains both runtime (**Unit Test**, **StringExpression Test**,**Validation Test**), and All table API/SQL **Integration Test**.

   **Third, the file name naming convention:**
* Unit Test -> functionName + **Test**
* StringExpression Test -> functionName + **StringExpressionTest**
* Validation Test -> functionName + **ValidationTest**
* Integration Test -> FunctionName + **ITCase**

**After the change of the directory structure, as shown below:**

image

@twalthr @fhueske @wuchong The change is big, so I need your help to ensure that the direction of change is correct.

Best,
SunJincheng

**Comment by wuchong** [comment]
Hi @sunjincheng121 , thanks for your great work. I like your proposal. The test structure refactoring looks good to me. I think it's good to put all the IT cases in one place.

**Comment by sunjincheng121** [comment]
Hi @wuchong Thanks for your approval. 
In addition to refactoring the test structure, I think that in addition to ITCast other tests should be unified use of `batchTestUtil` and` streamTestUtil`. What do you think? @wuchong @twalthr @fhueske

**Comment by sunjincheng121** [comment]
Rebase code...

**Comment by sunjincheng121** [comment]
May be I need split this PR into a couple of small PRs. What do you think? or have any other suggestions @fhueske  @wuchong @twalthr

**Comment by fhueske** [comment]
I think that is a very good idea. A review of a +6k -4k LOC PR takes a lot of time and has a good chance to be interrupted by other issues. Several smaller PRs would be much easier to review and merge. 
It would be great if you could do that @sunjincheng121.

Thank you very much!

**Comment by twalthr** [comment]
@sunjincheng121 the structure looks very good. I would volunteer to go over the changes as a whole and merge this. If @fhueske @wuchong are fine with this? Can you rebase the PR a last time?

**Comment by wuchong** [comment]
Thanks @twalthr , I'm fine with this ,a great +1!

**Comment by fhueske** [comment]
@twalthr, sure! Thanks for taking care of this PR.

**Comment by sunjincheng121** [comment]
@twalthr Extremely grateful If you can review and merge this PR after I rebase the PR.

**Comment by sunjincheng121** [comment]
@twalthr I have rebase the PR. I appreciated if you can review it. And i'll quick updated it when you left  comments.

Thanks,
SunJincheng

**Comment by sunjincheng121** [comment]
Update again. Because the new merged code.

**Comment by twalthr** [comment]
Thanks for the update @sunjincheng121. I will go over all changes and try to merge this.

**Comment by sunjincheng121** [comment]
Thanks @twalthr ! I have rebased the code and updated the PR. 

Thanks,
SunJincheng

**Comment by sunjincheng121** [comment]
Rebase code according master.

**Comment by twalthr** [comment]
@wuchong @sunjincheng121 @fhueske 

I tried to simplify the package structure a little bit. I based my changes on top of this PRs commit.

My goals:

- splitting a class like (`ExpressionReductionTest`) in 6 classes is not very helpful
- tests are named like the operator or feature that they are testing e.g. `Calc`, `Distinct` etc. (we don't need `CastingITCase`, `CalcITCase`, `ScalarFunctionITCase`; `CalcITCase` covers everything)
- all tests have the same name everywhere (`OverWindowTest`, `OverWindowValidationTest`, `OverWindowITCase`, `OverWindowStringExpressionTest`)
- no packages with one class
- `org.apache.flink.table.api` contains all tests that test the translation from batch/stream/table/sql APIs
- `org.apache.flink.table.plan` contains all tests that modify the plan
- `org.apache.flink.table.runtime` contains all ITCases and runtime relevant tests
- all other package contain unit tests for the classes of the same package

You can find my branch here: https://github.com/apache/flink/compare/master...twalthr:FLINK-6617

**Comment by fhueske** [comment]
Hi @twalthr and @sunjincheng121, I looked over the package structure and naming of the tests and this looks good to me.

+1 to merge from my side.

**Comment by sunjincheng121** [comment]
Hi @twalthr @fhueske thanks a lot for your review.
@twalthr your improve changes are make the structure of the package is clearer which make sense to me. 

+1 to merge.

**Comment by twalthr** [comment]
Merging...

**Pull request closed without merging**